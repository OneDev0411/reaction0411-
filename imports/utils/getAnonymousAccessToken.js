import Random from "@reactioncommerce/random";
import hashToken from "./hashToken.js";

/**
 * Generate a new secret token for the purpose of controlled access to
 * anonymous-user-specific data.
 *
 * @returns {Object} token object with associated token properties
 * @returns {Date} token.createdAt creation date, could potentially be useful for implementing expiration, but these do not currently expire.
 * @returns {String} token.hashedToken used to locate this token in the database
 * @returns {String} token.token raw token for use in URLs, email, etc. Do not store in the DB.
 */
export default function getAnonymousAccessToken() {
  const token = Random.secret();
  return {
    createdAt: new Date(),
    hashedToken: hashToken(token),
    token
  };
}
