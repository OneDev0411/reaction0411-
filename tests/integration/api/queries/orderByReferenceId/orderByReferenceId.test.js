import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import getAnonymousAccessToken from "@reactioncommerce/api-utils/getAnonymousAccessToken.js";
import Factory from "/tests/util/factory.js";
import TestApp from "/tests/util/TestApp.js";

const OrderByReferenceIdQuery = importAsString("./OrderByReferenceIdQuery.graphql");

jest.setTimeout(300000);

// Mock data for logged-in user test
const mockShopId = "integ-test-shop-id";
const opaqueShopId = "cmVhY3Rpb24vc2hvcDppbnRlZy10ZXN0LXNob3AtaWQ=";
const shopName = "Test Shop";
const orderId = "integ-test-order-id";
const orderReferenceId = "123456789";
const accountId = "integ-test-account-id";
const opaqueAccountId = "cmVhY3Rpb24vYWNjb3VudDppbnRlZy10ZXN0LWFjY291bnQtaWQ=";

const mockOrdersAccount = Factory.Account.makeOne({
  _id: accountId
});

const order = Factory.Order.makeOne({
  _id: orderId,
  referenceId: orderReferenceId,
  shopId: mockShopId,
  accountId: mockOrdersAccount._id
});

// Mock data for anonymous user order test
// const tokenInfo = hashToken(token);
const tokenInfo = getAnonymousAccessToken();
const orderIdAnon = "integ-test-order-id-anom";
const orderReferenceIdAnon = "123456789-anon";

const orderAnon = Factory.Order.makeOne({
  _id: orderIdAnon,
  referenceId: orderReferenceIdAnon,
  shopId: mockShopId,
  accountId: null,
  anonymousAccessTokens: [{ createdAt: tokenInfo.createdAt, hashedToken: tokenInfo.hashedToken }]
});

// Create test app
let testApp;
let query;

beforeAll(async () => {
  testApp = new TestApp();
  await testApp.start();
  query = testApp.query(OrderByReferenceIdQuery);
  await testApp.createUserAndAccount(mockOrdersAccount);
  await testApp.insertPrimaryShop({ _id: mockShopId, name: shopName });
});

beforeEach(async () => {
  await testApp.clearLoggedInUser();
  await testApp.collections.Orders.deleteMany({});
});

afterAll(async () => {
  await testApp.collections.Orders.deleteMany({});
  await testApp.collections.Shops.deleteMany({});
  await testApp.collections.users.deleteMany({});
  await testApp.collections.Accounts.deleteMany({});
  await testApp.stop();
});

test("get account order success", async () => {
  // Set up initial data state
  await testApp.collections.Orders.insertOne(order);
  await testApp.setLoggedInUser(mockOrdersAccount);

  // Query for order and check results
  const result = await query({ orderReferenceId, shopId: opaqueShopId, token: null });
  expect(result.orderByReferenceId.account._id).toBe(opaqueAccountId);
  expect(result.orderByReferenceId.shop._id).toBe(opaqueShopId);
  expect(result.orderByReferenceId.shop.name).toBe(shopName);
});

test("get account order failure", async () => {
  await testApp.collections.Orders.insertOne(order);
  await testApp.setLoggedInUser(mockOrdersAccount);
  const result = await query({ orderReferenceId: "does-not-exist", shopId: opaqueShopId, token: null });
  expect(result.orderByReferenceId).toBeNull();
});

test("get invalid params error", async () => {
  await testApp.collections.Orders.insertOne(orderAnon);
  try {
    await query({ orderReferenceId: orderReferenceIdAnon });
  } catch (error) {
    expect(error[0].message).toBe('Variable "$shopId" of required type "ID!" was not provided.');
  }
});

test("get anonymous order success", async () => {
  await testApp.collections.Orders.insertOne(orderAnon);
  const result = await query({ orderReferenceId: orderReferenceIdAnon, shopId: opaqueShopId, token: tokenInfo.token });
  expect(result.orderByReferenceId.shop._id).toBe(opaqueShopId);
  expect(result.orderByReferenceId.shop.name).toBe(shopName);
});
