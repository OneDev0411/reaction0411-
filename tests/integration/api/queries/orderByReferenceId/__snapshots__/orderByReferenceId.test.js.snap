// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get not found error for order that does not exist 1`] = `
Object {
  "extensions": Object {
    "code": "NOT_FOUND",
    "exception": Object {
      "details": Object {},
      "error": "not-found",
      "eventData": Object {},
      "isClientSafe": true,
      "reason": "Order not found",
    },
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "Order not found",
  "path": Array [
    "orderByReferenceId",
  ],
}
`;
